LibreNMS/Modules/Isis.php

```diff
@@ -0,0 +1,164 @@
+<?php
+/**
+ * Isis.php
+ *
+ * -Description-
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * @link       http://librenms.org
+ * @copyright  2021 Otto Reinikainen
+ * @author     Otto Reinikainen <otto@ottorei.fi>
+ */
+
+namespace LibreNMS\Modules;
+
+use App\Models\Device;
+use App\Models\IsisAdjacency;
+use LibreNMS\Interfaces\Module;
+use LibreNMS\OS;
+use LibreNMS\OS\Junos;
+use LibreNMS\Util\IP;
+
+class Isis implements Module
+{
+    /**
+     * Discover this module. Heavier processes can be run here
+     * Run infrequently (default 4 times a day)
+     *
+     * @param OS $os
+     */
+    public function discover(OS $os)
+    {
+        // not implemented
+    }
+
+    /**
+     * Poll data for this module and update the DB / RRD.
+     * Try to keep this efficient and only run if discovery has indicated there is a reason to run.
+     * Run frequently (default every 5 minutes)
+     *
+     * @param OS $os
+     */
+    public function poll(OS $os)
+    {
+        // Get device object
+        $device = $os->getDevice();
+        $device_id = $os->getDeviceId();
+
+        // Poll all ISIS enabled circuits from the device
+        $circuits_poll = $os->getCacheTable('ISIS-MIB::isisCirc', 'ISIS-MIB');
+
+        // Poll all available adjacencies
+        $adjacencies_poll = $os->getCacheTable('ISIS-MIB::isisISAdj', 'ISIS-MIB');
+        $adjacencies = collect();
+        $isis_data = [];
+
+        // Translate system state codes into meaningful strings
+        $isis_codes['1'] = 'L1';
+        $isis_codes['2'] = 'L2';
+        $isis_codes['3'] = 'L1L2';
+        $isis_codes['4'] = 'unknown';
+
+        if ($os instanceof Junos) {
+            // Do not poll loopback interface
+            unset($circuits_poll['16']);
+        }
+
+        // Loop through all configured adjacencies on the device
+        foreach ($circuits_poll as $circuit => $circuit_data) {
```

Comment:
Instead of this huge foreach loop, take a look at using the SyncsModels trait.  See https://github.com/librenms/librenms/blob/master/LibreNMS/Modules/Mempools.php#L57 for an example.

Reply:
All circuits would still need to be polled every time since the adjacency is lost when the configured adjacency goes down for a while. The only way to detect these is to first poll all configured circuits and then poll all adjacencies for these circuits. The circuit will always return the correct value if there is an ISIS-configuration on it but the adjacencies table does only return values if the adjacency is up at the moment the polling is done. That is why I am saving all values and updating them based on the ifIndex of the device. That way the missing and downed adjacencies can be updated and they also retaing useful information such as IfIndex and system id while they are down.

Reply:
After implementing SyncsModels in https://github.com/librenms/librenms/pull/12605, i can totally recommend it ^^